// @rxweb/reactive-form-validators/esm5/directives/template-validations/async-validation.directive.js
import * as tslib_1 from "tslib";
import { Input, Directive, forwardRef } from '@angular/core';
import { NG_ASYNC_VALIDATORS } from '@angular/forms';
import { of } from 'rxjs';
import * as ɵngcc0 from '@angular/core';
var AsyncValidationDirective = /** @class */ (function () {
    function AsyncValidationDirective() {
    }
    AsyncValidationDirective_1 = AsyncValidationDirective;
    AsyncValidationDirective.prototype.validate = function (control) {
        if (this.async)
            return this.async(control);
        return of(null);
    };
    var AsyncValidationDirective_1;
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], AsyncValidationDirective.prototype, "async", void 0);
AsyncValidationDirective.ɵfac = function AsyncValidationDirective_Factory(t) { return new (t || AsyncValidationDirective)(); };
AsyncValidationDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AsyncValidationDirective, selectors: [["", "ngModel", ""], ["", "formControlName", ""], ["", "formControl", ""]], inputs: { async: "async" }, features: [ɵngcc0.ɵɵProvidersFeature([{
                provide: NG_ASYNC_VALIDATORS,
                useExisting: forwardRef(function () { return AsyncValidationDirective_1; }),
                multi: true
            }])] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AsyncValidationDirective, [{
        type: Directive,
        args: [{
                selector: '[ngModel],[formControlName],[formControl]',
                providers: [{
                        provide: NG_ASYNC_VALIDATORS,
                        useExisting: forwardRef(function () { return AsyncValidationDirective_1; }),
                        multi: true
                    }]
            }]
    }], function () { return []; }, { async: [{
            type: Input
        }] }); })();
    return AsyncValidationDirective;
}());
export { AsyncValidationDirective };
