// @rxweb/reactive-form-validators/esm2015/directives/template-validations/async-validation.directive.js
import * as ɵngcc0 from '@angular/core';
var AsyncValidationDirective_1;
import * as tslib_1 from "tslib";
import { Input, Directive, forwardRef } from '@angular/core';
import { NG_ASYNC_VALIDATORS } from '@angular/forms';
import { of } from 'rxjs';
let AsyncValidationDirective = AsyncValidationDirective_1 = class AsyncValidationDirective {
    validate(control) {
        if (this.async)
            return this.async(control);
        return of(null);
    }
};
AsyncValidationDirective.ɵfac = function AsyncValidationDirective_Factory(t) { return new (t || AsyncValidationDirective)(); };
AsyncValidationDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: AsyncValidationDirective, selectors: [["", "ngModel", ""], ["", "formControlName", ""], ["", "formControl", ""]], inputs: { async: "async" }, features: [ɵngcc0.ɵɵProvidersFeature([{
                provide: NG_ASYNC_VALIDATORS,
                useExisting: forwardRef(() => AsyncValidationDirective_1),
                multi: true
            }])] });
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], AsyncValidationDirective.prototype, "async", void 0);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AsyncValidationDirective, [{
        type: Directive,
        args: [{
                selector: '[ngModel],[formControlName],[formControl]',
                providers: [{
                        provide: NG_ASYNC_VALIDATORS,
                        useExisting: forwardRef(() => AsyncValidationDirective_1),
                        multi: true
                    }]
            }]
    }], null, { async: [{
            type: Input
        }] }); })();
export { AsyncValidationDirective };
