// @rxweb/reactive-form-validators/esm5/directives/html-control-template.directive.js
import * as tslib_1 from "tslib";
import { Directive, TemplateRef, Input } from "@angular/core";
import * as ɵngcc0 from '@angular/core';
var HtmlControlTemplateDirective = /** @class */ (function () {
    function HtmlControlTemplateDirective(templateRef) {
        this.templateRef = templateRef;
    }
    ;
    tslib_1.__decorate([
        Input('htmlControlTemplate'),
        tslib_1.__metadata("design:type", String)
    ], HtmlControlTemplateDirective.prototype, "type", void 0);
    HtmlControlTemplateDirective = tslib_1.__decorate([ tslib_1.__metadata("design:paramtypes", [TemplateRef])
    ], HtmlControlTemplateDirective);
HtmlControlTemplateDirective.ɵfac = function HtmlControlTemplateDirective_Factory(t) { return new (t || HtmlControlTemplateDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
HtmlControlTemplateDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: HtmlControlTemplateDirective, selectors: [["", "htmlControlTemplate", ""]], inputs: { type: ["htmlControlTemplate", "type"] } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(HtmlControlTemplateDirective, [{
        type: Directive,
        args: [{
                selector: '[htmlControlTemplate]'
            }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }]; }, { type: [{
            type: Input,
            args: ['htmlControlTemplate']
        }] }); })();
    return HtmlControlTemplateDirective;
}());
export { HtmlControlTemplateDirective };
